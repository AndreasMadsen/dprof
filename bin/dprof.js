#!/bin/env node

var endpoint = require('endpoint');
var extend = require('util-extend');

process.stdin.pipe(endpoint(function (err, dprof) {
  if (err) throw err;
  dprof = JSON.parse(dprof);

  var content = new Node(dprof.root, dprof.total, 1, new State(), new Root());

  var svg = new Element('svg').attr({ xmlns: 'http://www.w3.org/2000/svg', width: 1000 });

  process.stdout.write(svg.start() + '\n');
  process.stdout.write(content.stringify() + '\n');
  process.stdout.write(svg.end());
}));

function Root() {
  this.top = 0;
}

function State() {
  this.y = 0;
}

function Node(node, total, depth, state, parent) {
  this.name = node.name;
  this.init = Math.floor(node.init / total * 1000);
  this.before = Math.floor((node.init + node.before) / total * 1000);
  this.after = Math.floor((node.init + node.after) / total * 1000);
  this.stack = node.stack;

  this.depth = depth;
  this.state = state;
  this.parent = parent;
  this.top = state.y;

  this.elems = [];
  this.elems.push(new Element('rect').attr({
    x: this.init,
    y: this.top,
    width: this.before - this.init,
    height: 20,
    fill: 'SteelBlue',
    style: 'shape-rendering: crispEdges'
  }));
  this.elems.push(new Element('rect').attr({
    x: this.before,
    y: this.top,
    width: this.after - this.before,
    height: 20,
    fill: 'IndianRed',
    style: 'shape-rendering: crispEdges'
  }));
  this.elems.push(new Element('line').attr({
    x1: this.init,
    y1: this.parent.top + 20,
    x2: this.init,
    y2: this.top + 20,
    style: 'stroke: #000; stroke-width: 2px'
  }));

  this.state.y += 20;
  this.children = node.children.map(function (child) {
    return new Node(child, total, depth + 1, state, this);
  }, this);
}

Node.prototype._indent = function () {
  var s = '';
  for (var i = 0; i < this.depth; i++) {
    s += '  ';
  }
  return s;
};

Node.prototype.stringify = function () {
  var content = [];
  this.elems.forEach(function (elem) {
    content.push(this._indent() + elem.singleton());
  }, this);
  this.children.forEach(function (child) {
    content.push(child.stringify());
  }, this);

  return content.join('\n');
};

function Element(name) {
  this._name = name;
  this._attr = {};
}

Element.prototype.attr = function (attr) {
  extend(this._attr, attr);
  return this;
};

Element.prototype.start = function () {
  var start = '<' + this._name;
  Object.keys(this._attr).forEach(function (k) {
    start += ' ' + k + '="' + this._attr[k] + '"';
  }, this);
  start += '>';
  return start;
};

Element.prototype.end = function () {
  return '</' + this._name + '>';
};

Element.prototype.singleton = function () {
  return this.start() + this.end();
};
