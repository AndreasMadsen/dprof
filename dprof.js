'use strict';

const providers = process.binding('async_wrap').Providers;
const asyncWrap = require('./async_wrap.js');
const zlib = require('zlib');
const fs = require('fs');

const version = require('./package.json').version;
const processStart = process.hrtime();

// Change the default stackTraceLimit if it haven't allready been overwritten
if (process.execArgv.indexOf('--stack_trace_limit') === -1 && Error.stackTraceLimit === 10) {
  Error.stackTraceLimit = 8;
}

//
// Define node class
//

function Site(site) {
  this.description = site.toString();
  this.filename = site.getFileName();
  this.line = site.getLineNumber();
  this.collum = site.getColumnNumber();
}

function timestamp() {
  const t = process.hrtime(processStart);
  return t[0] * 1e9 + t[1];
}

function Node(name, stack) {
  this.name = name;
  this._init = timestamp();
  this._before = [];
  this._after = [];
  this.children = [];
  this.stack = stack.map(function (site) {
    return new Site(site);
  });
}

Node.prototype.add = function (handle) {
  const node = new Node(handle.constructor.name, asyncWrap.stackTrace(3));
  this.children.push(node);
  return node;
};

Node.prototype.before = function () {
  this._before.push(timestamp());
};

Node.prototype.after = function () {
  this._after.push(timestamp());
};

Node.prototype.toJSON = function () {
  return {
    name: this.name,
    init: this._init,
    before: this._before,
    after: this._after,
    stack: this.stack,
    children: this.children
  };
};

Node.prototype.rootIntialize = function () {
  this._init = 0;
  this._before.push(0);
};

//
// Setup hooks
//

asyncWrap.setup(asyncInit, asyncBefore, asyncAfter);

const root = new Node('root', asyncWrap.stackTrace(2));
      root.rootIntialize();
let state = root;

function asyncInit(provider, parent) {
  if (provider === providers.TIMERWRAP) {
    this._dprofIgnore = true;
    return;
  }

  if (parent) {
    this._dprofState = parent._dprofState.add(this, provider, parent);
  } else {
    this._dprofState = state.add(this, provider, parent);
  }
}

function asyncBefore() {
  if (this._dprofIgnore) return;

  this._dprofState.before();
  state = this._dprofState;
}

function asyncAfter() {
  if (this._dprofIgnore) return;
  
  this._dprofState.after();
  state = root;
}

// The root job is done when process.nextTick is called
asyncWrap.disable();
process.nextTick(function () {
  root.after();
});
asyncWrap.enable();

//
// Print result
//

process.on('exit', function () {
  // even though zlib is sync, it still fires async_wrap events,
  // so disable asyncWrap just to be sure.
  asyncWrap.disable();

  const data = {
    'total': timestamp(),
    'version': version,
    'root': root
  };

  if (process.env.NODE_DPROF_DEBUG) {
    fs.writeFileSync('./dprof.json', JSON.stringify(data, null, 1));
  } else {
    fs.writeFileSync('./dprof.json.gz', zlib.gzipSync(JSON.stringify(data)));
  }
});
