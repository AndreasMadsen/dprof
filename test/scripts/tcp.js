'use strict';
require('../../dprof.js');

const net = require('net');

const server = net.createServer(function (socket) {
  socket.end('hallo world');
});

server.listen(0, 'localhost', function () {
  const addr = server.address();
  const socket = net.connect(addr.port, addr.host, function () {
    socket.once('readable', function () {
      socket.read();
      socket.once('readable', server.close.bind(server));
    });
  });
});
