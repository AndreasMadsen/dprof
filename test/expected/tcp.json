{
 "stdout": "",
 "stderr": "",
 "dump": {
  "total": 33357402,
  "version": "0.15.1",
  "root": {
   "name": "root",
   "parent": null,
   "uid": 0,
   "init": 0,
   "destroy": 12649140,
   "before": [
    0
   ],
   "after": [
    12470669
   ],
   "unrefed": false,
   "stack": [
    {
     "description": "Module._compile (module.js:541:32)",
     "filename": "module.js",
     "line": 541,
     "collum": 32
    },
    {
     "description": "Object.Module._extensions..js (module.js:550:10)",
     "filename": "module.js",
     "line": 550,
     "collum": 10
    },
    {
     "description": "Module.load (module.js:458:32)",
     "filename": "module.js",
     "line": 458,
     "collum": 32
    },
    {
     "description": "tryModuleLoad (module.js:417:12)",
     "filename": "module.js",
     "line": 417,
     "collum": 12
    },
    {
     "description": "Function.Module._load (module.js:409:3)",
     "filename": "module.js",
     "line": 409,
     "collum": 3
    },
    {
     "description": "Module.require (module.js:468:17)",
     "filename": "module.js",
     "line": 468,
     "collum": 17
    },
    {
     "description": "require (internal/module.js:20:19)",
     "filename": "internal/module.js",
     "line": 20,
     "collum": 19
    },
    {
     "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/tcp.js:2:1)",
     "filename": "/test/scripts/tcp.js",
     "line": 2,
     "collum": 1
    },
    {
     "description": "Module._compile (module.js:541:32)",
     "filename": "module.js",
     "line": 541,
     "collum": 32
    },
    {
     "description": "Object.Module._extensions..js (module.js:550:10)",
     "filename": "module.js",
     "line": 550,
     "collum": 10
    },
    {
     "description": "Module.load (module.js:458:32)",
     "filename": "module.js",
     "line": 458,
     "collum": 32
    },
    {
     "description": "tryModuleLoad (module.js:417:12)",
     "filename": "module.js",
     "line": 417,
     "collum": 12
    },
    {
     "description": "Function.Module._load (module.js:409:3)",
     "filename": "module.js",
     "line": 409,
     "collum": 3
    },
    {
     "description": "Module.runMain (module.js:575:10)",
     "filename": "module.js",
     "line": 575,
     "collum": 10
    },
    {
     "description": "run (bootstrap_node.js:352:7)",
     "filename": "bootstrap_node.js",
     "line": 352,
     "collum": 7
    },
    {
     "description": "startup (bootstrap_node.js:144:9)",
     "filename": "bootstrap_node.js",
     "line": 144,
     "collum": 9
    },
    {
     "description": "bootstrap_node.js:467:3",
     "filename": "bootstrap_node.js",
     "line": 467,
     "collum": 3
    }
   ],
   "children": [
    1,
    2
   ]
  },
  "nodes": [
   {
    "name": "Signal",
    "parent": 0,
    "uid": 1,
    "init": 2955615,
    "destroy": null,
    "before": [],
    "after": [],
    "unrefed": true,
    "stack": [
     {
      "description": "process.<anonymous> (internal/process.js:195:18)",
      "filename": "internal/process.js",
      "line": 195,
      "collum": 18
     },
     {
      "description": "emitTwo (events.js:106:13)",
      "filename": "events.js",
      "line": 106,
      "collum": 13
     },
     {
      "description": "process.emit (events.js:191:7)",
      "filename": "events.js",
      "line": 191,
      "collum": 7
     },
     {
      "description": "_addListener (events.js:226:14)",
      "filename": "events.js",
      "line": 226,
      "collum": 14
     },
     {
      "description": "process.addListener (events.js:270:10)",
      "filename": "events.js",
      "line": 270,
      "collum": 10
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/dprof.js:169:11)",
      "filename": "/dprof.js",
      "line": 169,
      "collum": 11
     },
     {
      "description": "Module._compile (module.js:541:32)",
      "filename": "module.js",
      "line": 541,
      "collum": 32
     },
     {
      "description": "Object.Module._extensions..js (module.js:550:10)",
      "filename": "module.js",
      "line": 550,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:458:32)",
      "filename": "module.js",
      "line": 458,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:417:12)",
      "filename": "module.js",
      "line": 417,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:409:3)",
      "filename": "module.js",
      "line": 409,
      "collum": 3
     },
     {
      "description": "Module.require (module.js:468:17)",
      "filename": "module.js",
      "line": 468,
      "collum": 17
     },
     {
      "description": "require (internal/module.js:20:19)",
      "filename": "internal/module.js",
      "line": 20,
      "collum": 19
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/tcp.js:2:1)",
      "filename": "/test/scripts/tcp.js",
      "line": 2,
      "collum": 1
     },
     {
      "description": "Module._compile (module.js:541:32)",
      "filename": "module.js",
      "line": 541,
      "collum": 32
     },
     {
      "description": "Object.Module._extensions..js (module.js:550:10)",
      "filename": "module.js",
      "line": 550,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:458:32)",
      "filename": "module.js",
      "line": 458,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:417:12)",
      "filename": "module.js",
      "line": 417,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:409:3)",
      "filename": "module.js",
      "line": 409,
      "collum": 3
     },
     {
      "description": "Module.runMain (module.js:575:10)",
      "filename": "module.js",
      "line": 575,
      "collum": 10
     },
     {
      "description": "run (bootstrap_node.js:352:7)",
      "filename": "bootstrap_node.js",
      "line": 352,
      "collum": 7
     },
     {
      "description": "startup (bootstrap_node.js:144:9)",
      "filename": "bootstrap_node.js",
      "line": 144,
      "collum": 9
     },
     {
      "description": "bootstrap_node.js:467:3",
      "filename": "bootstrap_node.js",
      "line": 467,
      "collum": 3
     }
    ],
    "children": []
   },
   {
    "name": "GetAddrInfoReqWrap",
    "parent": 0,
    "uid": 2,
    "init": 11839733,
    "destroy": 24533507,
    "before": [
     14196347
    ],
    "after": [
     21002773
    ],
    "unrefed": false,
    "stack": [
     {
      "description": "Object.lookup (dns.js:167:19)",
      "filename": "dns.js",
      "line": 167,
      "collum": 19
     },
     {
      "description": "listenAfterLookup (net.js:1393:20)",
      "filename": "net.js",
      "line": 1393,
      "collum": 20
     },
     {
      "description": "Server.listen (net.js:1389:5)",
      "filename": "net.js",
      "line": 1389,
      "collum": 5
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/tcp.js:10:8)",
      "filename": "/test/scripts/tcp.js",
      "line": 10,
      "collum": 8
     },
     {
      "description": "Module._compile (module.js:541:32)",
      "filename": "module.js",
      "line": 541,
      "collum": 32
     },
     {
      "description": "Object.Module._extensions..js (module.js:550:10)",
      "filename": "module.js",
      "line": 550,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:458:32)",
      "filename": "module.js",
      "line": 458,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:417:12)",
      "filename": "module.js",
      "line": 417,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:409:3)",
      "filename": "module.js",
      "line": 409,
      "collum": 3
     },
     {
      "description": "Module.runMain (module.js:575:10)",
      "filename": "module.js",
      "line": 575,
      "collum": 10
     },
     {
      "description": "run (bootstrap_node.js:352:7)",
      "filename": "bootstrap_node.js",
      "line": 352,
      "collum": 7
     },
     {
      "description": "startup (bootstrap_node.js:144:9)",
      "filename": "bootstrap_node.js",
      "line": 144,
      "collum": 9
     },
     {
      "description": "bootstrap_node.js:467:3",
      "filename": "bootstrap_node.js",
      "line": 467,
      "collum": 3
     }
    ],
    "children": [
     3,
     -1
    ]
   },
   {
    "name": "TCP",
    "parent": 2,
    "uid": 3,
    "init": 19691119,
    "destroy": 32559076,
    "before": [
     25666075
    ],
    "after": [
     27833127
    ],
    "unrefed": false,
    "stack": [
     {
      "description": "createServerHandle (net.js:1179:14)",
      "filename": "net.js",
      "line": 1179,
      "collum": 14
     },
     {
      "description": "Server._listen2 (net.js:1236:14)",
      "filename": "net.js",
      "line": 1236,
      "collum": 14
     },
     {
      "description": "listen (net.js:1288:10)",
      "filename": "net.js",
      "line": 1288,
      "collum": 10
     },
     {
      "description": "net.js:1398:9",
      "filename": "net.js",
      "line": 1398,
      "collum": 9
     },
     {
      "description": "GetAddrInfoReqWrap.asyncCallback [as callback] (dns.js:65:16)",
      "filename": "dns.js",
      "line": 65,
      "collum": 16
     },
     {
      "description": "GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:84:10)",
      "filename": "dns.js",
      "line": 84,
      "collum": 10
     }
    ],
    "children": [
     6,
     -4
    ]
   },
   {
    "name": "NextTickWrap",
    "parent": 2,
    "uid": -1,
    "init": 20823859,
    "destroy": 24114084,
    "before": [
     21141673
    ],
    "after": [
     24031618
    ],
    "unrefed": false,
    "stack": [
     {
      "description": "Server._listen2 (net.js:1266:11)",
      "filename": "net.js",
      "line": 1266,
      "collum": 11
     },
     {
      "description": "listen (net.js:1288:10)",
      "filename": "net.js",
      "line": 1288,
      "collum": 10
     },
     {
      "description": "net.js:1398:9",
      "filename": "net.js",
      "line": 1398,
      "collum": 9
     },
     {
      "description": "GetAddrInfoReqWrap.asyncCallback [as callback] (dns.js:65:16)",
      "filename": "dns.js",
      "line": 65,
      "collum": 16
     },
     {
      "description": "GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:84:10)",
      "filename": "dns.js",
      "line": 84,
      "collum": 10
     }
    ],
    "children": [
     4,
     -2
    ]
   },
   {
    "name": "TCP",
    "parent": -1,
    "uid": 4,
    "init": 23022423,
    "destroy": 32556881,
    "before": [
     28393736,
     29361280,
     32525569
    ],
    "after": [
     29285101,
     32272454,
     32553760
    ],
    "unrefed": false,
    "stack": [
     {
      "description": "Socket.connect (net.js:913:40)",
      "filename": "net.js",
      "line": 913,
      "collum": 40
     },
     {
      "description": "Object.exports.connect.exports.createConnection (net.js:70:35)",
      "filename": "net.js",
      "line": 70,
      "collum": 35
     },
     {
      "description": "Server.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/tcp.js:12:22)",
      "filename": "/test/scripts/tcp.js",
      "line": 12,
      "collum": 22
     },
     {
      "description": "Server.g (events.js:286:16)",
      "filename": "events.js",
      "line": 286,
      "collum": 16
     },
     {
      "description": "emitNone (events.js:86:13)",
      "filename": "events.js",
      "line": 86,
      "collum": 13
     },
     {
      "description": "Server.emit (events.js:185:7)",
      "filename": "events.js",
      "line": 185,
      "collum": 7
     },
     {
      "description": "emitListeningNT (net.js:1278:10)",
      "filename": "net.js",
      "line": 1278,
      "collum": 10
     },
     {
      "description": "_combinedTickCallback (internal/process/next_tick.js:71:11)",
      "filename": "internal/process/next_tick.js",
      "line": 71,
      "collum": 11
     },
     {
      "description": "process._tickCallback (internal/process/next_tick.js:98:9)",
      "filename": "internal/process/next_tick.js",
      "line": 98,
      "collum": 9
     }
    ],
    "children": [
     -5,
     -6,
     8
    ]
   },
   {
    "name": "NextTickWrap",
    "parent": -1,
    "uid": -2,
    "init": 23950962,
    "destroy": 24525072,
    "before": [
     24171290
    ],
    "after": [
     24510818
    ],
    "unrefed": false,
    "stack": [
     {
      "description": "lookupAndConnect (net.js:961:13)",
      "filename": "net.js",
      "line": 961,
      "collum": 13
     },
     {
      "description": "Socket.connect (net.js:929:5)",
      "filename": "net.js",
      "line": 929,
      "collum": 5
     },
     {
      "description": "Object.exports.connect.exports.createConnection (net.js:70:35)",
      "filename": "net.js",
      "line": 70,
      "collum": 35
     },
     {
      "description": "Server.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/tcp.js:12:22)",
      "filename": "/test/scripts/tcp.js",
      "line": 12,
      "collum": 22
     },
     {
      "description": "Server.g (events.js:286:16)",
      "filename": "events.js",
      "line": 286,
      "collum": 16
     },
     {
      "description": "emitNone (events.js:86:13)",
      "filename": "events.js",
      "line": 86,
      "collum": 13
     },
     {
      "description": "Server.emit (events.js:185:7)",
      "filename": "events.js",
      "line": 185,
      "collum": 7
     },
     {
      "description": "emitListeningNT (net.js:1278:10)",
      "filename": "net.js",
      "line": 1278,
      "collum": 10
     },
     {
      "description": "_combinedTickCallback (internal/process/next_tick.js:71:11)",
      "filename": "internal/process/next_tick.js",
      "line": 71,
      "collum": 11
     },
     {
      "description": "process._tickCallback (internal/process/next_tick.js:98:9)",
      "filename": "internal/process/next_tick.js",
      "line": 98,
      "collum": 9
     }
    ],
    "children": [
     5
    ]
   },
   {
    "name": "TCPConnectWrap",
    "parent": -2,
    "uid": 5,
    "init": 24394692,
    "destroy": 25570859,
    "before": [
     24617278
    ],
    "after": [
     25432397
    ],
    "unrefed": false,
    "stack": [
     {
      "description": "connect (net.js:854:26)",
      "filename": "net.js",
      "line": 854,
      "collum": 26
     },
     {
      "description": "net.js:963:9",
      "filename": "net.js",
      "line": 963,
      "collum": 9
     },
     {
      "description": "_combinedTickCallback (internal/process/next_tick.js:67:7)",
      "filename": "internal/process/next_tick.js",
      "line": 67,
      "collum": 7
     },
     {
      "description": "process._tickCallback (internal/process/next_tick.js:98:9)",
      "filename": "internal/process/next_tick.js",
      "line": 98,
      "collum": 9
     }
    ],
    "children": [
     -3
    ]
   },
   {
    "name": "NextTickWrap",
    "parent": 5,
    "uid": -3,
    "init": 24895141,
    "destroy": 25566167,
    "before": [
     25439373
    ],
    "after": [
     25562811
    ],
    "unrefed": false,
    "stack": [
     {
      "description": "Socket.Readable.on (_stream_readable.js:700:17)",
      "filename": "_stream_readable.js",
      "line": 700,
      "collum": 17
     },
     {
      "description": "Socket.once (events.js:296:8)",
      "filename": "events.js",
      "line": 296,
      "collum": 8
     },
     {
      "description": "Socket.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/tcp.js:13:12)",
      "filename": "/test/scripts/tcp.js",
      "line": 13,
      "collum": 12
     },
     {
      "description": "Socket.g (events.js:286:16)",
      "filename": "events.js",
      "line": 286,
      "collum": 16
     },
     {
      "description": "emitNone (events.js:86:13)",
      "filename": "events.js",
      "line": 86,
      "collum": 13
     },
     {
      "description": "Socket.emit (events.js:185:7)",
      "filename": "events.js",
      "line": 185,
      "collum": 7
     },
     {
      "description": "TCPConnectWrap.afterConnect [as oncomplete] (net.js:1067:10)",
      "filename": "net.js",
      "line": 1067,
      "collum": 10
     }
    ],
    "children": []
   },
   {
    "name": "TCP",
    "parent": 3,
    "uid": 6,
    "init": 25643251,
    "destroy": 33271280,
    "before": [
     32568838,
     33265678
    ],
    "after": [
     33141844,
     33269365
    ],
    "unrefed": false,
    "stack": [],
    "children": [
     -7,
     -8
    ]
   },
   {
    "name": "NextTickWrap",
    "parent": 3,
    "uid": -4,
    "init": 27548732,
    "destroy": 28194516,
    "before": [
     27844457
    ],
    "after": [
     28191758
    ],
    "unrefed": false,
    "stack": [
     {
      "description": "onwrite (_stream_writable.js:350:15)",
      "filename": "_stream_writable.js",
      "line": 350,
      "collum": 15
     },
     {
      "description": "WritableState.onwrite (_stream_writable.js:89:5)",
      "filename": "_stream_writable.js",
      "line": 89,
      "collum": 5
     },
     {
      "description": "Socket._writeGeneric (net.js:714:5)",
      "filename": "net.js",
      "line": 714,
      "collum": 5
     },
     {
      "description": "Socket._write (net.js:724:8)",
      "filename": "net.js",
      "line": 724,
      "collum": 8
     },
     {
      "description": "doWrite (_stream_writable.js:307:12)",
      "filename": "_stream_writable.js",
      "line": 307,
      "collum": 12
     },
     {
      "description": "writeOrBuffer (_stream_writable.js:293:5)",
      "filename": "_stream_writable.js",
      "line": 293,
      "collum": 5
     },
     {
      "description": "Socket.Writable.write (_stream_writable.js:220:11)",
      "filename": "_stream_writable.js",
      "line": 220,
      "collum": 11
     },
     {
      "description": "Socket.write (net.js:650:40)",
      "filename": "net.js",
      "line": 650,
      "collum": 40
     },
     {
      "description": "Socket.Writable.end (_stream_writable.js:453:10)",
      "filename": "_stream_writable.js",
      "line": 453,
      "collum": 10
     },
     {
      "description": "Socket.end (net.js:422:31)",
      "filename": "net.js",
      "line": 422,
      "collum": 31
     },
     {
      "description": "Server.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/tcp.js:7:10)",
      "filename": "/test/scripts/tcp.js",
      "line": 7,
      "collum": 10
     },
     {
      "description": "emitOne (events.js:96:13)",
      "filename": "events.js",
      "line": 96,
      "collum": 13
     },
     {
      "description": "Server.emit (events.js:188:7)",
      "filename": "events.js",
      "line": 188,
      "collum": 7
     },
     {
      "description": "TCP.onconnection (net.js:1458:8)",
      "filename": "net.js",
      "line": 1458,
      "collum": 8
     }
    ],
    "children": [
     7
    ]
   },
   {
    "name": "ShutdownWrap",
    "parent": -4,
    "uid": 7,
    "init": 28139332,
    "destroy": 28366378,
    "before": [
     28223881
    ],
    "after": [
     28362566
    ],
    "unrefed": false,
    "stack": [
     {
      "description": "Socket.onSocketFinish (net.js:225:26)",
      "filename": "net.js",
      "line": 225,
      "collum": 26
     },
     {
      "description": "emitNone (events.js:86:13)",
      "filename": "events.js",
      "line": 86,
      "collum": 13
     },
     {
      "description": "Socket.emit (events.js:185:7)",
      "filename": "events.js",
      "line": 185,
      "collum": 7
     },
     {
      "description": "finishMaybe (_stream_writable.js:488:14)",
      "filename": "_stream_writable.js",
      "line": 488,
      "collum": 14
     },
     {
      "description": "afterWrite (_stream_writable.js:362:3)",
      "filename": "_stream_writable.js",
      "line": 362,
      "collum": 3
     },
     {
      "description": "_combinedTickCallback (internal/process/next_tick.js:80:20)",
      "filename": "internal/process/next_tick.js",
      "line": 80,
      "collum": 20
     },
     {
      "description": "process._tickCallback (internal/process/next_tick.js:98:9)",
      "filename": "internal/process/next_tick.js",
      "line": 98,
      "collum": 9
     }
    ],
    "children": []
   },
   {
    "name": "NextTickWrap",
    "parent": 4,
    "uid": -5,
    "init": 29207279,
    "destroy": 29350420,
    "before": [
     29294645
    ],
    "after": [
     29347645
    ],
    "unrefed": false,
    "stack": [
     {
      "description": "maybeReadMore (_stream_readable.js:447:13)",
      "filename": "_stream_readable.js",
      "line": 447,
      "collum": 13
     },
     {
      "description": "readableAddChunk (_stream_readable.js:192:7)",
      "filename": "_stream_readable.js",
      "line": 192,
      "collum": 7
     },
     {
      "description": "Socket.Readable.push (_stream_readable.js:135:10)",
      "filename": "_stream_readable.js",
      "line": 135,
      "collum": 10
     },
     {
      "description": "TCP.onread (net.js:542:20)",
      "filename": "net.js",
      "line": 542,
      "collum": 20
     }
    ],
    "children": []
   },
   {
    "name": "NextTickWrap",
    "parent": 4,
    "uid": -6,
    "init": 29820379,
    "destroy": 32518842,
    "before": [
     32279819
    ],
    "after": [
     32514885
    ],
    "unrefed": false,
    "stack": [
     {
      "description": "endReadable (_stream_readable.js:966:13)",
      "filename": "_stream_readable.js",
      "line": 966,
      "collum": 13
     },
     {
      "description": "Socket.Readable.read (_stream_readable.js:300:7)",
      "filename": "_stream_readable.js",
      "line": 300,
      "collum": 7
     },
     {
      "description": "Socket.onSocketEnd (net.js:267:10)",
      "filename": "net.js",
      "line": 267,
      "collum": 10
     },
     {
      "description": "emitNone (events.js:86:13)",
      "filename": "events.js",
      "line": 86,
      "collum": 13
     },
     {
      "description": "Socket.emit (events.js:185:7)",
      "filename": "events.js",
      "line": 185,
      "collum": 7
     },
     {
      "description": "TCP.onread (net.js:579:8)",
      "filename": "net.js",
      "line": 579,
      "collum": 8
     }
    ],
    "children": []
   },
   {
    "name": "Pipe",
    "parent": 4,
    "uid": 8,
    "init": 30702093,
    "destroy": null,
    "before": [],
    "after": [],
    "unrefed": true,
    "stack": [
     {
      "description": "createHandle (net.js:33:31)",
      "filename": "net.js",
      "line": 33,
      "collum": 31
     },
     {
      "description": "new Socket (net.js:138:20)",
      "filename": "net.js",
      "line": 138,
      "collum": 20
     },
     {
      "description": "createWritableStdioStream (internal/process/stdio.js:156:16)",
      "filename": "internal/process/stdio.js",
      "line": 156,
      "collum": 16
     },
     {
      "description": "process.getStderr [as stderr] (internal/process/stdio.js:23:14)",
      "filename": "internal/process/stdio.js",
      "line": 23,
      "collum": 14
     },
     {
      "description": "Socket._destroy (net.js:484:25)",
      "filename": "net.js",
      "line": 484,
      "collum": 25
     },
     {
      "description": "Socket.destroy (net.js:518:8)",
      "filename": "net.js",
      "line": 518,
      "collum": 8
     },
     {
      "description": "Socket.onSocketFinish (net.js:213:17)",
      "filename": "net.js",
      "line": 213,
      "collum": 17
     },
     {
      "description": "emitNone (events.js:86:13)",
      "filename": "events.js",
      "line": 86,
      "collum": 13
     },
     {
      "description": "Socket.emit (events.js:185:7)",
      "filename": "events.js",
      "line": 185,
      "collum": 7
     },
     {
      "description": "finishMaybe (_stream_writable.js:488:14)",
      "filename": "_stream_writable.js",
      "line": 488,
      "collum": 14
     },
     {
      "description": "endWritable (_stream_writable.js:498:3)",
      "filename": "_stream_writable.js",
      "line": 498,
      "collum": 3
     },
     {
      "description": "Socket.Writable.end (_stream_writable.js:463:5)",
      "filename": "_stream_writable.js",
      "line": 463,
      "collum": 5
     },
     {
      "description": "Socket.end (net.js:422:31)",
      "filename": "net.js",
      "line": 422,
      "collum": 31
     },
     {
      "description": "Socket.destroySoon (net.js:449:10)",
      "filename": "net.js",
      "line": 449,
      "collum": 10
     },
     {
      "description": "Socket.onSocketEnd (net.js:272:10)",
      "filename": "net.js",
      "line": 272,
      "collum": 10
     },
     {
      "description": "emitNone (events.js:86:13)",
      "filename": "events.js",
      "line": 86,
      "collum": 13
     },
     {
      "description": "Socket.emit (events.js:185:7)",
      "filename": "events.js",
      "line": 185,
      "collum": 7
     },
     {
      "description": "TCP.onread (net.js:579:8)",
      "filename": "net.js",
      "line": 579,
      "collum": 8
     }
    ],
    "children": []
   },
   {
    "name": "NextTickWrap",
    "parent": 6,
    "uid": -7,
    "init": 32875796,
    "destroy": 33184285,
    "before": [
     33145696
    ],
    "after": [
     33181889
    ],
    "unrefed": false,
    "stack": [
     {
      "description": "Server._emitCloseIfDrained (net.js:1542:11)",
      "filename": "net.js",
      "line": 1542,
      "collum": 11
     },
     {
      "description": "Socket._destroy (net.js:510:20)",
      "filename": "net.js",
      "line": 510,
      "collum": 20
     },
     {
      "description": "Socket.destroy (net.js:518:8)",
      "filename": "net.js",
      "line": 518,
      "collum": 8
     },
     {
      "description": "maybeDestroy (net.js:442:12)",
      "filename": "net.js",
      "line": 442,
      "collum": 12
     },
     {
      "description": "TCP.onread (net.js:570:5)",
      "filename": "net.js",
      "line": 570,
      "collum": 5
     }
    ],
    "children": []
   },
   {
    "name": "NextTickWrap",
    "parent": 6,
    "uid": -8,
    "init": 33042578,
    "destroy": 33261038,
    "before": [
     33186177
    ],
    "after": [
     33258611
    ],
    "unrefed": false,
    "stack": [
     {
      "description": "endReadable (_stream_readable.js:966:13)",
      "filename": "_stream_readable.js",
      "line": 966,
      "collum": 13
     },
     {
      "description": "Socket.Readable.read (_stream_readable.js:300:7)",
      "filename": "_stream_readable.js",
      "line": 300,
      "collum": 7
     },
     {
      "description": "Socket.read (net.js:299:43)",
      "filename": "net.js",
      "line": 299,
      "collum": 43
     },
     {
      "description": "Socket.onSocketEnd (net.js:267:10)",
      "filename": "net.js",
      "line": 267,
      "collum": 10
     },
     {
      "description": "emitNone (events.js:91:20)",
      "filename": "events.js",
      "line": 91,
      "collum": 20
     },
     {
      "description": "Socket.emit (events.js:185:7)",
      "filename": "events.js",
      "line": 185,
      "collum": 7
     },
     {
      "description": "TCP.onread (net.js:579:8)",
      "filename": "net.js",
      "line": 579,
      "collum": 8
     }
    ],
    "children": []
   }
  ]
 }
}